File: src/amenities/amenities_server/application/Services.cs
using amenities_server.persistence;
using amenities_server.services;
using amenities_server.validators;
using Microsoft.Extensions.DependencyInjection;

namespace amenities_server.application
{
    public static class Services
    {
        private static IAmenityPersistence? _amenityPersistence = null;
        private static IPermissionValidator? _permissionValidator = null;

        /*
        Returns a singleton of the amenity persistence, constructs one if it does not exist
        Args:
            None
        Returns:
            IAmenityPersistence: The singleton instance of AmenityPersistence
        */
        public static IAmenityPersistence GetAmenityPersistence()
        {
            //return a singleton of the amenity persistence, construct one if it does not exist
            _amenityPersistence ??= ConstructAmenityPersistence();
            return _amenityPersistence;
        }

        /*
        Returns a singleton of the permission validator, constructs one if it does not exist
        Args:
            IHttpClientFactory httpFactory: A factory to create HTTP clients for permission validation
        Returns:
            IPermissionValidator: The singleton instance of PermissionValidator
        */
        public static IPermissionValidator GetPermissionValidator(IHttpClientFactory httpFactory)
        {
            //return a singleton of the permission validator, construct one if it does not exist
            _permissionValidator ??= ConstructPermissionValidator(httpFactory);
            return _permissionValidator;
        }

        /*
        Sets the AmenityPersistence instance
        Args:
            IAmenityPersistence amenityPersistence: The instance of AmenityPersistence to set
        Returns:
            void
        */
        public static void SetAmenityPersistence(IAmenityPersistence amenityPersistence)
        {
            _amenityPersistence = amenityPersistence;
        }

        /*
        Clears the stored AmenityPersistence instance
        Args:
            None
        Returns:
            void
        */
        public static void Clear()
        {
            _amenityPersistence = null;
        }

        /*
        Constructs the AmenityPersistence instance based on environment variables
        Args:
            None
        Returns:
            IAmenityPersistence: The constructed instance of AmenityPersistence
        */
        private static IAmenityPersistence ConstructAmenityPersistence()
        {
            //construct a instance of the amenity persistence, creates a mock or postgres implementation based on the env variable 'DB_IMPLEMENTATION' stated in the yaml file
            string? dbImplementation = Environment.GetEnvironmentVariable("DB_IMPLEMENTATION") ?? string.Empty;
            IAmenityPersistence? amenityPersistence = null;
            if (dbImplementation == "POSTGRES")
            {
                Console.WriteLine("Attempting to connect to Postgres");
                try
                {
                    amenityPersistence = new PostgresAmenityPersistence(PostgresConnectionString());
                    Console.WriteLine("Postgress Connected successfully");
                }
                catch (InvalidOperationException)
                {
                    amenityPersistence = null;
                    Console.WriteLine("Postgress Failed");
                }
            }
            else if (dbImplementation == "MOCK")
            {
                amenityPersistence = new StubAmenityPersistence();
            }

            return amenityPersistence ?? new StubAmenityPersistence();
        }

        /*
        Constructs the PermissionValidator instance
        Args:
            IHttpClientFactory httpFactory: A factory to create HTTP clients for permission validation
        Returns:
            IPermissionValidator: The constructed instance of PermissionValidator
        */
        private static IPermissionValidator ConstructPermissionValidator(IHttpClientFactory httpFactory)
        {
            Console.WriteLine("Constructing PermissionValidator");
            var PermCli = new PermissionClient(httpFactory);
            return new PermissionValidator(PermCli);
        }

        /*
        Returns the Postgres connection string using environment variables
        Args:
            None
        Returns:
            string: The constructed Postgres connection string
        */
        private static string PostgresConnectionString()
        {
            //if a postgres connection is to be made, grab connection details from the environment as stated in the yaml file
            string? dbHost = Environment.GetEnvironmentVariable("DB_HOST") ?? string.Empty;
            string? dbPort = Environment.GetEnvironmentVariable("DB_PORT") ?? string.Empty;
            string? dbUsername = Environment.GetEnvironmentVariable("DB_USERNAME") ?? string.Empty;
            string? dbPassword = Environment.GetEnvironmentVariable("DB_PASSWORD") ?? string.Empty;
            return $"Host={dbHost}; Port={dbPort}; Username={dbUsername}; Password={dbPassword}";
        }
    }
}

File: src/incident_reports/incident_reports_server/application/services.py
from incident_reports_server.validators.permission_validator import PermissionValidator
from incident_reports_server.persistence.i_incident_report_persistence import IIncidentReportPersistence
from incident_reports_server.persistence.incident_report_persistence_mongo import IncidentReportPersistenceMongo
from incident_reports_server.persistence.incident_report_persistence_stub import IncidentReportPersistenceStub
import os

class Services:
    _incident_report_persistence = None 
    _permission_validator = None
    
    @staticmethod
    def get_incident_report_persistence() -> IIncidentReportPersistence:
        """
        Retrieves the singleton instance of the Incident Report Persistence class.
        If it does not exist, it constructs it using the appropriate persistence layer (MongoDB or Mock).
        
        Returns:
            IIncidentReportPersistence: The persistence instance (either MongoDB or Mock).
        """
        if Services._incident_report_persistence is None:
            Services._incident_report_persistence = Services._construct_incident_report_persistence()
            
        return Services._incident_report_persistence 
    
    @staticmethod
    def get_permission_validator():
        """
        Retrieves the singleton instance of the Permission Validator.
        If it does not exist, it constructs it using the default constructor.
        
        Returns:
            PermissionValidator: The permission validator instance.
        """
        if Services._permission_validator is None:
            Services._permission_validator = Services._construct_permission_validator()
        
        return Services._permission_validator
    
    @staticmethod
    def _construct_incident_report_persistence() -> IIncidentReportPersistence:
        """
        Constructs the appropriate persistence instance based on the DB_IMPLEMENTATION environment variable.
        If MongoDB is specified, it attempts to establish a connection using the provided connection string.
        If Mock is specified, it uses a mock/stub implementation.
        
        Returns:
            IIncidentReportPersistence: The constructed persistence instance (either MongoDB or Mock).
        """
        # Construct a persistence instance, whether it constructs mock or a mongodb connection depends on the env variable 'DB_IMPLEMENTATION' initialized by the docker file
        db_implementation = os.getenv("DB_IMPLEMENTATION") 
        
        if db_implementation == "MONGODB":
            """
            Attempt to connect to MongoDB using the connection string.
            """
            print(f"Attempting to connect to MongoDB with connection string: {Services.db_connection_string()}")
            try:
                # Initialize MongoDB persistence with the connection string and DB name
                db_implementation = IncidentReportPersistenceMongo(Services.db_connection_string())
                print("Successfully connected to MongoDB")
            except Exception as e:
                # If the connection fails, log the exception and use a stub instead
                db_implementation = None
                print(f'MongoDB failed to connect: {e}')
        
        elif db_implementation == "MOCK":
            """
            Use a mock/stub implementation for testing or alternative environments.
            """
            db_implementation = IncidentReportPersistenceStub()
        
        if not db_implementation:
            db_implementation = IncidentReportPersistenceStub()
        
        return db_implementation 
    
    @staticmethod
    def _construct_permission_validator():
        """
        Constructs and returns the PermissionValidator instance.
        
        Returns:
            PermissionValidator: The constructed permission validator instance.
        """
        return PermissionValidator()
    
    @staticmethod
    def db_connection_string() -> str:
        """
        Constructs the MongoDB connection string using environment variables (DB_HOST, DB_PORT, DB_USERNAME, DB_PASSWORD).
        This string is used to connect to the MongoDB instance for persistent storage.
        
        Returns:
            str: The MongoDB connection string.
        """
        # If a mongodb connection is to be established, pull connection details from the docker file
        mongo_host = os.getenv('DB_HOST', 'mongo')
        mongo_port = int(os.getenv('DB_PORT', 27017))
        mongo_username = os.getenv('DB_USERNAME', 'mongo_db_user')
        mongo_password = os.getenv('DB_PASSWORD', 'password')
        
        # Format the connection string
        return f'mongodb://{mongo_username}:{mongo_password}@{mongo_host}:{mongo_port}'

    @staticmethod
    def clear() -> None:
        """
        Clears the singleton instance of the Incident Report Persistence and releases any resources.
        This is useful for cleaning up and resetting the service state.
        """
        # If instance exists, clear the data and release the singleton
        if Services._incident_report_persistence:
            Services._incident_report_persistence.clear()  # Clear the persistence data
            Services._incident_report_persistence = None
